phys/module_diag_solar.F

```diff
@@ -252,78 +255,99 @@ SUBROUTINE solar_diag (rho, dz8w, ph, phb, cldfrac3d, coszen, swdnb, swdnt, qv,
                  else
                    tau_qs(i, j) = 0.0
                  end if
-               else
-                 re_snow_path(i, j) = MISSING
-                 tau_qs(i, j) = MISSING
                end if
+             else  
+               swp2d(i, j) = MISSING
+               re_snow_path(i, j) = MISSING
+               tau_qs(i, j) = MISSING
              end if
 
-             if (f_qc .or. f_qi .or. f_qs) then
+             if ( (f_qc .or. f_qi .or. f_qs) .and. & 
+                  (p_qc > param_first_scalar .or. &
+                   p_qi > param_first_scalar .or. &
+                   p_qs > param_first_scalar) ) then
                  !!! RESOLVED !!!
                  ! Sum water paths (cloud liquid + ice + snow)
-               wp2d_sum(i, j) = lwp2d(i, j) + iwp2d(i, j) + swp2d(i, j)
+               wp2d_sum(i, j) = MAX( lwp2d(i, j), 0.0 ) + MAX( iwp2d(i, j), 0.0 ) + MAX( swp2d(i, j), 0.0 )
 
                  !!! CLOUD BASE & TOP HEIGHTS
                  ! Cloud base first
-               k = kts
-               wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)
-               do while ( (wc < WC_MIN) .and. (k < kte) )
-                 k = k + 1
+               if (wp2d_sum(i, j) > Q_MIN) then
+                 k = kts
                  wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)
-               end do
+                 do while ( (wc < WC_MIN) .and. (k < kte) )
+                   k = k + 1
+                   wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)
```

Comment:
... but here we assume that they all are available
wc = qc(i, k, j) + qi(i, k, j) + qs(i, k, j)

Has this been tested with a simple MP scheme, such as WSM3?

Reply:
@davegill  Thanks for pointing this out. You are right -- this does appear to be an inconsistency. I just tested with WSM3, and variable wc is calculated correctly (i.e., variables qi and qs are equal to 0.0 such that wc = qc). Results were the same when compiling in debug mode. I can fix the code to be more explicit about which hydrometeor classes are considered if you would like. Thanks.

Reply:
@davegill @twjuliano Actually when you use WSM3, qi and qs are not in the memory. So this code is probably not safe.

Reply:
@twjuliano 
Tim,
We definitely need some IF tests here. Just above, you use f_qc and f_qi (and p_qc and p_qi). Those existence tests should be part of this assignment.

We are hoping to push the release-v4.2.2 out next week. No pressure.

Reply:
@weiwangncar 
Wei,
We need to have some logic in place to safeguard against use of variables that are not allocated. Even if things appear OK, this is still a concern.

Reply:
@twjuliano 
Tim, 
Wow, that extra logic for qi and qs does really muddy up the ease of understanding! 

This change looks to be exactly what we were asking for. 

In the comments below, would you state that you have run one of your test cases with this new logic and that the results are still OK.

Reply:
@davegill 
Great. I've edited my latest comment at the bottom.